rclcpp/test/rclcpp/test_generic_pubsub.cpp

```diff
@@ -147,14 +171,59 @@ TEST_F(RclcppGenericNodeFixture, publisher_and_subscriber_work)
   ASSERT_TRUE(success);
 
   for (const auto & message : test_messages) {
-    publisher->publish(serialize_string_message(message));
+    publisher->publish(serialize_message<std::string, test_msgs::msg::Strings>(message));
   }
 
   auto subscribed_messages = subscriber_future_.get();
   EXPECT_THAT(subscribed_messages, SizeIs(Not(0)));
   EXPECT_THAT(subscribed_messages[0], StrEq("Hello World"));
 }
 
+TEST_F(RclcppGenericNodeFixture, publish_loaned_msg_work)
+{
+  // We currently publish more messages because they can get lost
+  std::vector<int64_t> test_messages = {100, 100};
+  std::string topic_name = "/int64_topic";
+  std::string type = "test_msgs/msg/BasicTypes";
+
+  auto publisher = node_->create_generic_publisher(topic_name, type, rclcpp::QoS(1));
+
+  if (publisher->can_loan_messages()) {
+    auto subscriber_future_ = std::async(
+      std::launch::deferred, [this, topic_name, type] {
+        return subscribe_raw_messages<int64_t, test_msgs::msg::BasicTypes>(
+          1, topic_name, type);
+      });
+
+    auto allocator = node_->get_node_options().allocator();
+    auto success = false;
+    auto ret = rcl_wait_for_subscribers(
+      node_->get_node_base_interface()->get_rcl_node_handle(),
+      &allocator,
+      topic_name.c_str(),
+      1u,
+      static_cast<rcutils_duration_value_t>(2e9),
+      &success);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ASSERT_TRUE(success);
+
+    for (const auto & message : test_messages) {
+      publisher->publish_loaned_msg(
+        serialize_message<int64_t, test_msgs::msg::BasicTypes>(message));
+    }
+
+    auto subscribed_messages = subscriber_future_.get();
+    EXPECT_THAT(subscribed_messages, SizeIs(Not(0)));
+    EXPECT_EQ(subscribed_messages[0], test_messages[0]);
+  } else {
+    ASSERT_THROW(
+    {
+      publisher->publish_loaned_msg(
+        serialize_message<int64_t, test_msgs::msg::BasicTypes>(test_messages[0]));
+    }, rclcpp::exceptions::RCLError);
```

Comment:
Can it be tested? BTW, Do you think we need to add the failure case for `publish_loaned_msg` with `test_msgs::msg::Strings`

Reply:
> Can it be tested?  
Yes. Currently we should do configuration to enable shared memory feature for cyclonedds and fastdds.
By default, test always execute this code.   
For the success case for calling publish_loaned_message, we have to do test in local environment (different dds need to different configuration).